plugins/modules/zos_stat.py

```diff
@@ -0,0 +1,2157 @@
+    else:
+        return GenerationDataGroupHandler(name, volume, module, sms_managed, ds_type, gdg_view, tmp_hlq)
+
+
+def get_facts_handler(
+    name,
+    resource_type,
+    module,
+    volume=None,
+    tmp_hlq=None,
+    sms_managed=False,
+    file_args=None
+):
+    """Returns the correct handler needed depending on the type of resource
+    we will query.
+
+    Arguments:
+        name (str) -- Name of the resource.
+        resource_type (str) -- One of 'data_set', 'aggregate' or 'file'.
+        module (AnsibleModule) -- Ansible object with the task's context.
+        volume (str, optional) -- Volume where a data set is allocated.
+        tmp_hlq (str, optional) -- Temp HLQ for certain data set operations.
+        sms_managed (bool, optional) -- Whether a data set is managed by SMS.
+        file_args (dict, optional) -- Options affecting how a file is query.
+
+    Returns:
+        FactsHandler: Handler for data sets/aggregates/files.
+    """
+    if resource_type == 'data_set':
+        return get_data_set_handler(name, volume, module, tmp_hlq, sms_managed)
+    elif resource_type == 'file':
+        return FileHandler(name, module, file_args)
+    elif resource_type == 'aggregate':
+        return AggregateHandler(name, module)
+
+
+def run_module():
+    """Parses the module's options and retrieves facts accordingly.
+    """
+    module = AnsibleModule(
+        argument_spec={
+            'name': {
+                'type': 'str',
+                'required': True,
+                'aliases': ['src']
+            },
+            'volume': {
+                'type': 'str',
+                'required': False
+            },
+            'type': {
+                'type': 'str',
+                'required': False,
+                'default': 'data_set',
+                'choices': ['data_set', 'file', 'aggregate']
+            },
+            'sms_managed': {
+                'type': 'bool',
+                'required': False,
+                'default': False
+            },
+            'tmp_hlq': {
+                'type': 'str',
+                'required': False
+            },
+            'follow': {
+                'type': 'bool',
+                'required': False,
+                'default': False
+            },
+            'get_mime': {
+                'type': 'bool',
+                'required': False,
+                'default': True
+            },
+            'get_checksum': {
+                'type': 'bool',
+                'required': False,
+                'default': True
+            },
+            'checksum_algorithm': {
+                'type': 'str',
+                'required': False,
+                'default': 'sha1',
+                'choices': ['md5', 'sha1', 'sha224', 'sha256', 'sha384', 'sha512']
+            }
+        },
+        required_if=[
+            # Forcing a volume when querying data sets.
+            ('type', 'data_set', ('volume',))
```

Comment:
Was forcing a volume decided in the design phase of the module? I would think normal operation would be, we use the volume in the master catalog if the data set is cataloged.

Reply:
Yes, since not forcing the volume would entail first finding the data set, at which point the module starts to duplicate the functionality of zos_find. The expected workflow if a user doesn't know which volume a data set resides in, is to first use zos_find and then zos_stat.